src/number_theory/modular_forms/basic.lean

```diff
@@ -0,0 +1,311 @@
+
+instance has_sub : has_sub (modular_form Γ k) :=
+⟨ λ f g, f + -g ⟩
+
+@[simp] lemma coe_sub (f g : (modular_form Γ k)) : ⇑(f - g) = f - g := rfl
+
+@[simp] lemma sub_apply (f g : modular_form Γ k) (z : ℍ) : (f - g) z = f z - g z := rfl
+
+instance : add_comm_group (modular_form Γ k) :=
+fun_like.coe_injective.add_comm_group _ rfl coe_add coe_neg coe_sub coe_smul coe_smul
+
+lemma coe_zero : ⇑(0 : modular_form Γ k) = (0 : ℍ → ℂ) := rfl
+
+/--Additive coercion from `modular_form` to `ℍ → ℂ`. -/
+def coe_hom : (modular_form Γ k) →+ (ℍ → ℂ) :=
+{ to_fun := λ f, f,
+  map_zero' := coe_zero,
+  map_add' := λ _ _, rfl }
+
+lemma coe_hom_injective : function.injective (@coe_hom Γ k) :=
+fun_like.coe_injective
+
+instance : module ℂ (modular_form Γ k) :=
+coe_hom_injective.module ℂ (coe_hom) (λ _ _, rfl)
+
+instance : inhabited (modular_form Γ k) := ⟨0⟩
+
+/--The modular form of weight `k_1 + k_2` given by the product of two modular forms of weights
+`k_1` and `k_2`. -/
+def mul {k_1 k_2 : ℤ} {Γ : subgroup SL(2, ℤ)} (f : (modular_form Γ k_1))
+  (g : (modular_form Γ k_2)) : (modular_form Γ (k_1 + k_2)) :=
+{ to_fun := f * g,
+  slash_action_eq' := λ A, by simp_rw [mul_slash_subgroup, modular_form_class.slash_action_eq],
+  hol' := f.hol'.mul g.hol',
+  bdd_at_infty' := λ A, begin
+    rw [mul_slash_SL2],
+    exact (f.bdd_at_infty' A).mul (g.bdd_at_infty' A),
+  end}
+
+  /-- The constant function is bounded at infinity. -/
+lemma is_bounded_at_im_infty_one : is_bounded_at_im_infty (1 : ℍ → ℂ):=
+@asymptotics.is_O_const_const _ _ ℂ _ _ 1 _ one_ne_zero _
+
+instance : has_one (modular_form Γ 0) :=
+{ one :=
+    { hol' := λ x, mdifferentiable_at_const 𝓘(ℂ, ℂ) 𝓘(ℂ, ℂ),
+    bdd_at_infty' := λ A, (congr_arg is_bounded_at_im_infty (is_invariant_one A)).mpr
+      is_bounded_at_im_infty_one,
+    .. (1 : slash_invariant_form Γ 0) } }
+
+end modular_form
+
+namespace cusp_form
+
+variables {F : Type*} {Γ : subgroup SL(2, ℤ)} {k : ℤ}
+
+instance has_add : has_add (cusp_form Γ k) :=
+{ add := λ f g,
+  { to_fun := f + g,
+    hol' := f.hol'.add g.hol',
+    zero_at_infty' := λ A, by simpa using (zero_at_im_infty_submodule ℂ).add_mem'
+      (f.zero_at_infty' A) (g.zero_at_infty' A),
+    .. (f : slash_invariant_form Γ k) + g }}
+
+@[simp] lemma coe_add (f g : cusp_form Γ k) : ⇑(f + g) = f + g := rfl
+
+@[simp] lemma add_apply (f g : cusp_form Γ k) (z : ℍ) : (f + g) z = f z + g z := rfl
+
+instance has_zero : has_zero (cusp_form Γ k) :=
+⟨ { to_fun := 0,
+    slash_action_eq' := slash_action.mul_zero _,
+    hol' := (λ _, mdifferentiable_at_const 𝓘(ℂ, ℂ) 𝓘(ℂ, ℂ)),
+    zero_at_infty' := λ A, by simpa [slash_action.mul_zero _ ]
+      using filter.zero_is_zero_at_filter _, }⟩
+
+section
+variables {α : Type*} [has_smul α ℂ] [is_scalar_tower α ℂ ℂ]
+
+instance has_smul : has_smul α (cusp_form Γ k) :=
+⟨ λ c f,
+  { to_fun := c • f,
+    hol' := by simpa using f.hol'.const_smul (c • (1 : ℂ)),
```

Comment:
I think this indicates that `f.hol'.const_smul` should be more general

Reply:
Sure, maybe we can do it in or after #17793 ?